W-11862525-first-draft-for-beta-site-build-CP (#114)
mulesoft/docs-apikit: modules/ROOT/pages/configure-headers4-task.adoc

> = Configuring Response Headers and Status Code

Reviewer: ```suggestion
= Configuring the Response Headers and Status Code
```

Author: OK! Thanks!

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Within the HTTP Listener configuration, you have the option to customize the HTTP response body, headers, status code and reason phrase.

Reviewer: ```suggestion
Within the HTTP Listener configuration, you can customize the HTTP response body, headers, status code, and reason phrase.
```

Author: OK! Thanks!

== Configure Variable Names

> The APIkit Router configuration encompasses two specific properties that enable you to configure the variables that the HTTP listener searches for by default:

Reviewer: ```suggestion
The APIkit Router configuration has two properties that enable you to configure the variables that the HTTP listener searches for by default:
```

Author: OK! Thanks!

> * `outboundHeadersMapName`: Name of the variable that contains a headers map added in the response (`outboundHeaders` by default).

Reviewer: Remove periods for these bullets. They are not full sentences.

Author: OK! Thanks!

+

> * `httpStatusVarName`: Name of the variable that stores the status code to be set in the response (`httpStatus` by default).

Reviewer: ```suggestion
* `httpStatusVarName`: Name of the variable that stores the status code to set in the response (`httpStatus` by default).
```

Author: OK! Thanks!

Both variables are created by the APIkit Router and facilitate the inclusion of custom HTTP headers and status codes in the response.

// TODO Maybe an image of the Router configuration as well?

[source,xml,linenums]
----
<apikit:config  name="api-config"
                api="api.raml"
                outboundHeadersMapName="outboundHeaders"
                httpStatusVarName="httpStatus"/>
----

> If you opt to modify the default names of these variables, it is crucial to update the default HTTP listener configuration so to ensure alignment between the variable names and the configuration.

Reviewer: ```suggestion
If you opt to modify the default names of these variables, it is crucial to update the default HTTP listener configuration so to ensure alignment between the variable names and the configuration.
```
```suggestion
To modify the default names of these variables, it is crucial to update the default HTTP listener configuration to ensure alignment between the variable names and the configuration.
```

Author: OK! Thanks!

[source,xml,linenums]
----
<http:listener config-ref="api-httpListenerConfig" path="/api/*">
    <http:response
        statusCode="#[vars.httpStatus default 200]">
        <http:headers>#[vars.outboundHeaders default {}]</http:headers>
    </http:response>
    <http:error-response
        statusCode="#[vars.httpStatus default 500]">
        <http:body>#[payload]</http:body>
        <http:headers>#[vars.outboundHeaders default {}]</http:headers>
    </http:error-response>
</http:listener>
----

== Add a Header to a Response

Put the header you want to add inside the `outboundHeaders` map, as shown in the following example:

[source,xml,linenums]
----
<set-variable
    value='#[vars.outboundHeaders default {} ++ {newHeaderName: "Header value"}]'
    variableName="outboundHeaders"/>
----

== Set the Status Code of a Response

To assign a specific value to the `httpStatus` variable, you can use the following example as a reference:

[source,xml,linenums]
----
<set-variable
    value="201"
    variableName="httpStatus"/>
----

== See Also
// TODO Fix the reference to the HTTP listener

> * https://docs.mulesoft.com/http-connector/1.7/http-listener-ref#sending-http-responses[Sending HTTP Responses when configuring HTTP listener source in Mule 4^]

Reviewer: This text should be intial capped if it is a topic title.  Also, the carat is for use when we link outside of docs.mulesoft.com.

Author: OK! Thanks!

> = Configuring the Response Headers and Status Code

Reviewer 2: In this case, gerund is correct - there are multiple tasks (imperative) in this file.

Author: @hannanelson, thanks for your feedback!
I understand that we should use gerund in H1 and infinitive in H2+.
Actually, the APIkit beta site (https://dev-docs-internal.kqa.msap.io/beta-apikit-new-content-structure/apikit/4.x/overview) has been structured like that.
Please, can you confirm if gerund is OK for H1 and infinitive for H2+?
Thanks!

Reviewer 2: Yes, if there are multiple tasks in a file.

Author: In this file, there is more than one task, so I understand that it is OK to use the structure gerund for H1 and infinitive for H2. 
Thanks for your support!

Reviewer 2: Yes.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

In the *HTTP Listener* configuration, you can customize the HTTP response body, headers, status code, and reason phrase.

== Configure Variable Names

> You can configure the variables that the HTTP listener searches for by default by configuring the following two properties in *APIkit Router*:

Reviewer 2: ```suggestion
You can configure the default variables that the HTTP listener searches for by configuring these properties in *APIkit Router*:
```

Reviewer 2: (not sure if that changes the meaning, but the "by default" sounds awkward in the original.)

Author: The "by default by" sounds weird!
Thanks for solving it! :)

* `outboundHeadersMapName`
+
Name of the variable that contains a headers map added in the response (`outboundHeaders` by default)

* `httpStatusVarName`
+
Name of the variable that stores the status code to set in the response (`httpStatus` by default)

> *APIkit Router* creates both variables and facilitates the inclusion of custom HTTP headers and status codes in the response.

Reviewer 2: ```suggestion
*APIkit Router* creates both variables and facilitates including custom HTTP headers and status codes in the response.
```

Author: OK! Thanks!

// TODO: Add an image of the Router configuration

[source,xml,linenums]
----
<apikit:config  name="api-config"
                api="api.raml"
                outboundHeadersMapName="outboundHeaders"
                httpStatusVarName="httpStatus"/>
----

To modify the default names of these variables, update the default HTTP listener configuration to align the variable names and the configuration.

[source,xml,linenums]
----
<http:listener config-ref="api-httpListenerConfig" path="/api/*">
    <http:response
        statusCode="#[vars.httpStatus default 200]">
        <http:headers>#[vars.outboundHeaders default {}]</http:headers>
    </http:response>
    <http:error-response
        statusCode="#[vars.httpStatus default 500]">
        <http:body>#[payload]</http:body>
        <http:headers>#[vars.outboundHeaders default {}]</http:headers>
    </http:error-response>
</http:listener>
----

== Add a Header to a Response

> Put the header to add in the `outboundHeaders` map, as shown in the following example:

Reviewer 2: Nice tightening!

Author: Thanks!
